composer/core/data_spec.py

```diff
@@ -18,6 +18,68 @@
 __all__ = ["DataSpec"]
 
 
+def _split_list(l, num_microbatches):
+    if len(l) < num_microbatches:
+        raise ValueError(
+            textwrap.dedent(f"""\
+        Cannot split list of length {len(l)} into {num_microbatches} batches.
+         make sure `grad_accum` is less than or equal to `train_batch_size // world_size`."""))
+    return [l[i::num_microbatches] for i in range(num_microbatches)]
+
+
+def _split_tensor(t, num_microbatches):
+    if len(t) < num_microbatches:
+        raise ValueError(
+            textwrap.dedent(f"""\
+        Cannot split tensor of length {len(t)} into {num_microbatches} batches.
+         make sure `grad_accum` is less than or equal to `train_batch_size // world_size`."""))
+    return t.chunk(num_microbatches)
+
+
+def _split_mapping(m, num_microbatches):
+    chunked = {}
+    for k, v in m.items():
+        if isinstance(v, torch.Tensor):
+            chunked[k] = _split_tensor(v, num_microbatches)
+        if isinstance(v, (List, Tuple)):
+            chunked[k] = _split_list(v, num_microbatches)
+    return [{k: v[idx] for k, v in chunked.items()} for idx in range(num_microbatches)]
+
+
+def _default_split_batch(batch: Any, num_microbatches: int) -> Sequence:
```

Comment:
Would more specific typing work (where `TBatch = TypeVar('TBatch')`)?
```suggestion
def _default_split_batch(batch: TBatch, num_microbatches: int) -> Sequence[TBatch]:
```